go_pprof_scraper/datadog_checks/go_pprof_scraper/check.py

```diff
@@ -0,0 +1,208 @@
+# (C) Datadog, Inc. 2022-present
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+import datetime
+import time
+from concurrent.futures import ThreadPoolExecutor
+from uuid import uuid4
+
+import requests_unixsocket
+from requests.exceptions import ConnectionError, HTTPError, InvalidURL, Timeout
+
+from datadog_checks.base import AgentCheck, ConfigurationError
+
+try:
+    from urllib.parse import quote, urljoin, urlparse
+except ImportError:
+    # Use Python 2 imports
+    from urllib import quote
+
+    from urlparse import urljoin, urlparse
+
+
+try:
+    import datadog_agent
+except ImportError:
+    from datadog_checks.stubs import datadog_agent
+
+
+VALID_PROFILE_TYPES = {"cpu", "heap", "mutex", "block", "goroutine"}
+DEFAULT_PROFILE_DURATION = 30  # TODO: something more sensible
+DEFAULT_PROFILES = ["cpu", "heap"]
+
+
+def _timeformat(dt):
+    return dt.strftime("%Y-%m-%dT%H:%M:%SZ")
+
+
+class GoPprofScraperCheck(AgentCheck):
+
+    # This will be the prefix of every metric and service check the integration sends
+    __NAMESPACE__ = "go_pprof_scraper"
+
+    def __init__(self, name, init_config, instances):
+        super(GoPprofScraperCheck, self).__init__(name, init_config, instances)
+
+        self._get_apm_config()
+        self.runtime_id = uuid4()
+
+        self.service = self.instance.get("service_name")
+        if not self.service:
+            raise ConfigurationError("service_name is required")
+        self.env = self.instance.get("env", datadog_agent.get_config("env"))
+
+        self.url = self.instance.get("pprof_url")
+        if not self.url:
+            raise ConfigurationError("pprof_url is required")
+        self.hostname = urlparse(self.url).hostname
+        # TODO: What if self.hostname == "localhost" (or 127.0.0.1 or whatever?)
+
+        self.duration = self.instance.get("duration", DEFAULT_PROFILE_DURATION)
+        if self.duration > self.instance.get("min_collection_interval", 60):
+            raise ConfigurationError(
+                "duration ({}) is be longer than min_collection_interval ({})".format(
+                    self.duration, self.instance.get("min_collection_interval")
+                )
+            )
+
+        self.profiles = self.instance.get("profiles", DEFAULT_PROFILES)
+        for profile in self.profiles:
+            if profile not in VALID_PROFILE_TYPES:
+                raise ConfigurationError(
+                    "{} is not a valid profile type, must be one of {}".format(profile, VALID_PROFILE_TYPES)
+                )
+        self.profiles = list(set(self.profiles))
+
+        self.cumulative = self.instance.get("cumulative", True)
+        self.tags = self.instance.get("tags", [])
+        self.tags.extend(self.init_config.get("tags", []))
+
+    def _get_apm_config(self):
+        # TODO: for some reason the config I get here doesn't match the real
```

Comment:
This is probably the biggest issue I've found so far. If I start the agent without explicitly enabling APM, The `datadog_agent.get_config("apm_config")` call will still return a blob of configuration that says APM is enabled. But then the check subsequently fails to upload profiles to the profile proxy since it's not actually running.

Reply:
[This](https://github.com/DataDog/integrations-extras/pull/1541/files#diff-6321a81caa84fe8dfa36d5a3c1be0d1d118eeaebfd4fda66a530b2334c060da2R11-R18) is the config you get when running the test env because we use [this](https://github.com/DataDog/integrations-extras/pull/1541/files#diff-6321a81caa84fe8dfa36d5a3c1be0d1d118eeaebfd4fda66a530b2334c060da2R22-R37) to start it

Reply:
@ofek any idea?

Reply:
I've poked at this a little more, and I'm not sure now where the `datadog_agent.get_config("apm_config")` call is actually getting its config _from_. I tried setting the [`DD_APM_RECEIVER_PORT`](https://github.com/DataDog/datadog-agent/blob/e4e52fff0a85928725e436d6fe092dfa7856b974/pkg/config/config_template.yaml#L1067) env var to something other than the default, and the agent started with that value but `get_config` didn't notice it. Specifically, in [`conftest.py`](https://github.com/DataDog/integrations-extras/pull/1541/files#diff-6321a81caa84fe8dfa36d5a3c1be0d1d118eeaebfd4fda66a530b2334c060da2R27) I set the variable to 9876. Then I added a log line to print the `get_config` value. Then when I start the testing environment I see these two logs:

```
2022-11-06 12:56:26 UTC | TRACE | INFO | (run.go:254 in Infof) | Listening for traces at http://0.0.0.0:9876
...
2022-11-06 12:56:39 UTC | CORE | INFO | (pkg/collector/python/datadog_agent.go:127 in LogMessage) | unknown | (check.py:94) | apm_config: {'enabled': True, 'receiver_port': 8126, 'remote_tagger': True, 'telemetry': {'enabled': True}, 'windows_pipe_buffer_size': 1000000, 'windows_pipe_security_descriptor': 'D:AI(A;;GA;;;WD)'}
```

So the agent is getting configured the way I expect, but the check isn't seeing the correct configuration. The same thing happens if I don't set `DD_APM_ENABLED=true`: the trace agent (where profiles go) won't actually be enabled, but the configuration I read says it is.

Reply:
To be clear, this isn't necessarily a blocker, assuming most people just use the default configuration. But I can't think of any way to work around this in case someone doesn't use the defaults.